src/contexts/auth-provider.tsx

```diff
@@ -39,39 +42,61 @@ const AuthProvider: React.FC<{children: ReactNode}> = ({ children }) => {
   const dispatch = useAppDispatch();
   const [solveAuthChallenge] = useSolveChallengeMutation();
   const [requestAuthChallenge] = useRequestChallengeMutation();
+  const [refreshTokenMutation] = useRefreshTokenMutation();
+
   const token = cookies.get("bearerToken");
   const { publicKey, signMessage, disconnect } = useWallet();
   const [address, setAddress] = useState('');
+  const { data: usdcBalanceData, refetch: refetchUsdcBalance } = useGetUserUsdcBalanceQuery({ user: address }, { skip: !address });
 
-  const validateToken = useCallback((challengeResponse: ChallengeResponse) => {
-    if (!publicKey) return;
-
-    const decoded = jwt_decode<JwtPayload>(challengeResponse.token);
-    const address = publicKey.toBase58();
-    if (decoded.sub !== address) {
-      cookies.remove('bearerToken');
-      dispatch(setLoginStatus(LoginStatus.OUT));
-      return;
-    }
+  const resetAuth = useCallback(async () => {
+    cookies.remove('bearerToken');
+    await disconnect();
+    setAddress('');
+    dispatch(setUsdcBalance(0));
+    dispatch(setLoginStatus(LoginStatus.OUT));
+  }, [disconnect, cookies, dispatch]);
 
-    if (decoded.exp && Date.now() >= decoded.exp * 1000) {
-      console.warn('Token expired. Please refresh.');
-      cookies.remove('bearerToken');
-      dispatch(setLoginStatus(LoginStatus.OUT));
-      return;
+  const handleTokenValidation = useCallback(async (token: string, address: string) => {
+    try {
+      const decoded = jwt_decode<JwtPayload>(token);
+      
+      if (decoded.sub !== address) {
+        resetAuth();
+        return false;
+      }
+  
+      if (decoded.exp && Date.now() >= decoded.exp * 1000) {
+        const refreshedToken = await refreshTokenMutation({ token }).unwrap();
+        token = refreshedToken.token;
+      }
+  
+      // Set the cookie here
+      cookies.set('bearerToken', token, {
+        path: '/',
+        maxAge: decoded.exp ? decoded.exp * 1000 - Date.now() : undefined,
+        expires: decoded.exp ? new Date(decoded.exp * 1000) : undefined,
+        secure: true,
+        sameSite: 'strict'
+      });
+  
+      setAddress(address);
+      dispatch(setLoginStatus(LoginStatus.IN));
+      LogRocket.identify(address);
+  
+      // Delay the refetch to ensure the address has been set
+      setTimeout(() => {
+        refetchUsdcBalance();
+      }, 100);
+  
+      return true;
+    } catch (error) {
+      console.error("Failed to validate or refresh token", error);
+      resetAuth();
+      return false;
     }
-
-    cookies.set('bearerToken', challengeResponse.token, {
-      path: '/',
-      maxAge: challengeResponse.valid_til,
-      expires: new Date(challengeResponse.valid_til),
-      secure: true,
-    });
-    setAddress(address);
-    dispatch(getTransactions({ address }));
-    LogRocket.identify(address);
```

Comment:
Can we continue to use LogRocket? Will be helpful for identifying user struggles